crates/typst/src/layout/grid/layout.rs

```diff
@@ -411,204 +429,311 @@ impl CellGrid {
-                        // The current value of the auto index is always the
-                        // index of the latest automatically positioned cell
-                        // placed plus one (that's what we do in
-                        // 'resolve_cell_position'), so we subtract 1 to get
-                        // that cell's index, and place the vline after its
-                        // column. The exception is when the auto_index is 0,
-                        // meaning no automatically positioned cell was placed
-                        // yet. In that case, we place the vline to the left of
-                        // the table.
-                        auto_index
-                            .checked_sub(1)
-                            .map_or(0, |last_auto_index| last_auto_index % c + 1)
-                    });
-                    if end.is_some_and(|end| end.get() < start) {
-                        bail!(span, "line cannot end before it starts");
-                    }
-                    let line = Line { index: x, start, end, stroke, position };
 
-                    // For consistency with hlines, we only push vlines to the
-                    // final vector of vlines after processing every cell.
-                    pending_vlines.push((span, line));
-                    continue;
-                }
-                GridItem::Cell(cell) => cell,
-            };
-            let cell_span = cell.span();
-            // Let's calculate the cell's final position based on its
-            // requested position.
-            let resolved_index = {
-                let cell_x = cell.x(styles);
-                let cell_y = cell.y(styles);
-                resolve_cell_position(cell_x, cell_y, &resolved_cells, &mut auto_index, c)
-                    .at(cell_span)?
-            };
-            let x = resolved_index % c;
-            let y = resolved_index / c;
-            let colspan = cell.colspan(styles).get();
-            let rowspan = cell.rowspan(styles).get();
+                    is_header = true;
+                    header_span = span;
+                    repeat_header = repeat;
 
-            if colspan > c - x {
-                bail!(
-                    cell_span,
-                    "cell's colspan would cause it to exceed the available column(s)";
-                    hint: "try placing the cell in another position or reducing its colspan"
-                )
-            }
+                    // If any cell in the header is automatically positioned,
+                    // have it skip to the next row. This is to avoid having a
+                    // header after a partially filled row just add cells to
+                    // that row instead of starting a new one.
+                    // FIXME: Revise this approach when headers can start from
+                    // arbitrary rows.
+                    min_auto_index = auto_index.next_multiple_of(c);
```

Comment:
why does just `auto_index = auto_index.next_multiple_of(c);` not work?

Reply:
The idea is to not advance it if you don't place an automatically positioned cell in the header. Later on we ensure the auto index is at least after the header after we finish writing the header, but that doesn't mean we change it (it might be after the last row in the header). This system will be made more robust once headers can start from arbitrary rows, but I believe this should work for now.